### PR TITLE
Download all sources for `dist-git-source`

### PR DESCRIPTION
**Pull request by lukaszachy**
Document common options on a single place
Option to just download sources (no extractions)
Fix 'url' behaviour for 'shell'
Move 'distgit_download' to utils

Pull Request Checklist

* [x] implement the feature
* [x] write the documentation
* [x] extend the test coverage
* [x] update the specification
* [x] adjust plugin docstring
* [x] modify the json schema
* [x] mention the version

### CODE REVIEW

**Comment by lukaszachy** [comment]
I'll do changes to 'how: fmf' and tests once I know whether this is a step in the correct direction.

Example:

```
$ cat demo.fmf
discover:
  how: shell
  dist-git-source: true
  dist-git-download-only: true
  url: https://gitlab.com/redhat/centos-stream/rpms/ruby.git
execute:
  how: tmt
```

For `tmt run --id DEMO discover` one should have  all necessary files in var/tmp/tmt/DEMO/demo/discover/default-0/source to run `rpmbuild -bp`

**Comment by lukaszachy** [comment]
Note to self: `TMT_SOURCE_DIR` is not available in the PREPARE steps.

File: spec/plans/discover.fmf [hunk 1]
```diff
@@ -236,6 +230,55 @@ description: |
...
+    summary: Download rpm sources
+    description: |
+      There is a support to download sources of the rpm.
+      Plan using the option has to be defined in DistGit(Fedora, CentOS)
+      or ``url`` option needs to point top such repository.
+      At this moment no patches are applied, only sources are downloaded
+      and tarballs extracted.
+
+      All source files are available for further use in directory stored in
+      ``TMT_SOURCE_DIR`` variable.
```
Comment (lukaszachy) [inline spec/plans/discover.fmf:243]:
```suggestion
      ``TMT_SOURCE_DIR`` variable available in tests.
```

**Comment by lukaszachy** [comment]
Q about the option name. The standard-test-role uses 'fetch_only': Should we call it the same? 

https://pagure.io/standard-test-roles/blob/master/f/roles/standard-test-source

**Review by CtrlZmaster — COMMENTED** [review]
There were a couple of places where the sentences didn't make sense to me, so I suggested changes. In other places, I felt like that the text could be interpreted incorrectly by the users, so I tried to make it more clear.

Regardless of my suggestions, this massively improved the documentation and implemented what was missing for my use case.  Thank you!

File: spec/plans/discover.fmf [hunk 2]
```diff
@@ -185,20 +172,27 @@ description: |
...
+            files.
+
+        dist-git-remove-fmf-root
+            Set to True to remove any fmf root present in the sources.
+            ``dist-git-init`` can be used to create it later at desired location.
+
+        dist-git-merge
+            Set to True to combine fmf root from the sources and fmf root from plan.
+            For example when ``dist-git-download-only`` is used this option allows to
+            find tests defined within plans repo.
```
[the remark this answers is not in the document]
Comment (lukaszachy) [reply spec/plans/discover.fmf:190]: It is all just about tests. Any plans in extracted sources will not be processed. The 'current plan' which asks for the dist-git-source is already being executed (and remote plans are processed before the discovery step starts).
I'll use your suggestion as a base.

File: tmt/steps/discover/__init__.py [hunk 3]
```diff
@@ -37,6 +37,13 @@ class DiscoverStepData(tmt.steps.WhereableStepData, tmt.steps.StepData):
         help='Use the provided DistGit handler instead of the auto detection.'
         )
 
+    dist_git_fetch_only: bool = field(
```
[the remark this answers is not in the document]
Comment (lukaszachy) [reply tmt/steps/discover/__init__.py:40]: Yeah, however standard-test-roles uses fetch-only. Hard to tell what is better

File: spec/plans/discover.fmf (hunk 2 above)
[the remark this answers is not in the document]
Comment (lukaszachy) [reply spec/plans/discover.fmf:190]: How about:
```
        dist-git-merge
            Set to True to combine fmf root from the sources and fmf root from the plan.
            It allows to have tests defined in the DistGit repo which use resources
            from the extracted sources.
```

File: tmt/steps/discover/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (lukaszachy) [reply tmt/steps/discover/__init__.py:40]: Will rename to download-only.

File: spec/plans/discover.fmf (hunk 2 above)
[the remark this answers is not in the document]
Comment (CtrlZmaster) [reply spec/plans/discover.fmf]: In my opinion, the comment and the suggestion are not telling the same thing, or the suggestion is being too general. The word resources is very abstract and I wouldn't really understand that it also include tests. I understood what you wrote in the first comment, so I incorporated it and I think that now it is much clearer. 

```
 dist-git-merge
             Set to True to combine fmf root from the sources and fmf root from the plan.
             It allows to have plans and tests defined in the DistGit repo which use tests and 
             other resources from the downloaded sources. Any plans in extracted sources 
             will not be processed.
```

**Comment by lukaszachy** [comment]
Hopefully all comments  addressed now.  Approvals appreciated :)

**Review by psss — CHANGES_REQUESTED** [review]
Looks good and the wording sounds much more clear now. Proposing a couple of adjustments in 3fd2bb36 and added a few questions.

File: tests/discover/distgit.sh [hunk 4]
```diff
@@ -59,7 +63,7 @@ rlJournalStart
         rlRun "popd"
     rlPhaseEnd
 
-
+if false; then
```
[the remark this answers is not in the document]
Comment (lukaszachy) [reply tests/discover/distgit.sh:66]: Ouch

**Comment by lukaszachy** [comment]
@psss Thanks for finding those issues, fixed in [f281bb6](https://github.com/teemtee/tmt/pull/2438/commits/f281bb62f016810af1d41b5dd10b1533e0d0c4ca)

**Review by psss — APPROVED** [review]
Now looks good. Thanks for this nice improvement!

**Pull request merged**